### PR TITLE
[ML] Validate streaming HTTP Response

### PR DESCRIPTION
**Pull request by prwhelan**
Read the first HttpResult from the stream and validate the HttpResponse object, invoking the listener to retry or fail the request.  On success, ResponseHandlers will create an InferenceServiceResults with the Flow.Publisher to parse the bytes for the provider.

### CODE REVIEW

**Comment by prwhelan** [comment]
@elasticmachine update branch

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/StreamingResponseHandler.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        }
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        if (upstreamIsClosed.compareAndSet(false, true)) {
+            if (downstream != null) {
+                downstream.onError(throwable);
+            } else {
+                log.debug(
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/StreamingResponseHandler.java:122]:
```suggestion
                log.warn(
```

Looks like something unexpected happened

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java [hunk 2]
```diff
@@ -108,23 +108,29 @@ public void tryAction(ActionListener<InferenceServiceResults> listener) {
...
             });
 
             try {
-                httpClient.send(request.createHttpRequest(), context, responseListener);
+                if (request.isStreaming() && responseHandler.canHandleStreamingResponses()) {
+                    httpClient.stream(request.createHttpRequest(), context, retryableListener.delegateFailure((l, r) -> {
+                        r.subscribe(new StreamingResponseHandler(throttlerManager, logger, request, responseHandler, l));
+                    }));
+                } else {
+                    httpClient.send(request.createHttpRequest(), context, retryableListener.delegateFailure((l, r) -> {
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java:122]: If is adds complications then I can make a case for not supporting retry with streaming completions.

Retry is useful for an unattended bulk operation such as embeddings at ingest, it's better to retry and overcome transient issues. Streaming completions is attended, someone is likely reading the response if they want to retry they can press the button again. Same with a client.

**Review by davidkyle — APPROVED** [review]
LGTM

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/StreamingResponseHandler.java (hunk 1 above)
Comment (prwhelan) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/StreamingResponseHandler.java:122]: Do we have to use the `throttleManager` for that?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java (hunk 2 above)
Comment (prwhelan) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java:122]: We would really only process the retry before the connection and stream are established - if the connection is interrupted mid-stream or something, we currently do not retry, we'll just forward the error to the user

**Comment by prwhelan** [comment]
@elasticmachine update branch

**Pull request merged**